// introcs/list/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Node = (function () {
    function Node(value, next) {
        this.value = value;
        this.next = next;
    }
    Node.prototype.toString = function () {
        var result = "" + this.value + " → ";
        if (this.next === null) {
            result += "null";
        }
        else {
            result += this.next;
        }
        return result;
    };
    return Node;
}());
exports.Node = Node;
exports.first = function (list) {
    if (list === null) {
        throw new Error("Cannot call function `first` on null.");
    }
    else {
        return list.value;
    }
};
exports.rest = function (list) {
    if (list === null) {
        throw new Error("Cannot call function `rest` on null.");
    }
    else {
        return list.next;
    }
};
exports.cons = function (value, list) {
    if (list === void 0) { list = null; }
    return new Node(value, list);
};
exports.listify = function () {
    var values = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        values[_i] = arguments[_i];
    }
    if (values.length === 0) {
        return null;
    }
    else {
        return exports.cons(values[0], exports.listify.apply(null, values.slice(1)));
    }
};
