// introcs/turtle/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./style.css");
var TurtleProvider_1 = require("./TurtleProvider");
var ForwardCommand_1 = require("./commands/ForwardCommand");
var TurnCommand_1 = require("./commands/TurnCommand");
var MoveToCommand_1 = require("./commands/MoveToCommand");
var TurnToCommand_1 = require("./commands/TurnToCommand");
var SpeedCommand_1 = require("./commands/SpeedCommand");
var BackwardCommand_1 = require("./commands/BackwardCommand");
var StrokeCommand_1 = require("./commands/StrokeCommand");
exports.forward = function (amount) {
    TurtleProvider_1.TurtleProvider.instance().emit(new ForwardCommand_1.ForwardCommand(amount));
};
exports.backward = function (amount) {
    TurtleProvider_1.TurtleProvider.instance().emit(new BackwardCommand_1.BackwardCommand(amount));
};
exports.left = function (angle) {
    TurtleProvider_1.TurtleProvider.instance().emit(new TurnCommand_1.TurnCommand(angle));
};
exports.right = function (angle) {
    TurtleProvider_1.TurtleProvider.instance().emit(new TurnCommand_1.TurnCommand(-angle));
};
exports.moveTo = function (x, y) {
    TurtleProvider_1.TurtleProvider.instance().emit(new MoveToCommand_1.MoveToCommand(x, y));
};
exports.turnTo = function (angle) {
    TurtleProvider_1.TurtleProvider.instance().emit(new TurnToCommand_1.TurnToCommand(angle));
};
exports.setStroke = function (stroke) {
    TurtleProvider_1.TurtleProvider.instance().emit(new StrokeCommand_1.StrokeCommand(stroke));
};
exports.setSpeed = function (speed) {
    TurtleProvider_1.TurtleProvider.instance().emit(new SpeedCommand_1.SpeedCommand(speed));
};
exports.setOnClick = function (handler) {
    TurtleProvider_1.TurtleProvider.instance().setOnClick(handler);
};
