// introcs/graphics/Transform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Transform = (function () {
    function Transform() {
        this.a = 1;
        this.b = 0;
        this.c = 0;
        this.d = 1;
        this.e = 0;
        this.f = 0;
    }
    Transform.prototype.copy = function () {
        var t = new Transform();
        t.a = this.a;
        t.b = this.b;
        t.c = this.c;
        t.d = this.d;
        t.e = this.e;
        t.f = this.f;
        return t;
    };
    Transform.prototype.translate = function (x, y) {
        var t = this.copy();
        t.e += x;
        if (y) {
            t.f += y;
        }
        return t;
    };
    Transform.prototype.scale = function (x, y) {
        var t = this.copy();
        t.a *= x;
        if (y === undefined) {
            t.d *= x;
        }
        else {
            t.d *= y;
        }
        return t;
    };
    Transform.prototype.rotate = function (a) {
        var t = this.copy();
        var aBase = 0;
        if (t.a !== 0) {
            aBase = Math.acos(t.a);
        }
        var aHat = aBase + a;
        t.a = Math.cos(aHat);
        t.b = Math.sin(aHat);
        t.c = -Math.sin(aHat);
        t.d = Math.cos(aHat);
        return t;
    };
    Transform.prototype.toMatrix = function () {
        return "matrix(" + this.a + "," + this.b + "," + this.c + "," + this.d + "," + this.e + "," + this.f + ")";
    };
    Transform.DEFAULT = new Transform();
    return Transform;
}());
exports.default = Transform;
