// introcs/console/DOM.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function element(tag, className, children) {
    var element = document.createElement(tag);
    if (Array.isArray(children)) {
        for (var _i = 0, children_1 = children; _i < children_1.length; _i++) {
            var child = children_1[_i];
            element.appendChild(child);
        }
    }
    else if (children !== null) {
        element.innerText = children;
    }
    element.setAttribute("class", className);
    return element;
}
function div(className, children) {
    return element("div", className, children);
}
exports.div = div;
function p(className, children) {
    return element("p", className, children);
}
exports.p = p;
function table(className, children) {
    return element("table", className, children);
}
exports.table = table;
function thead(className, children) {
    return element("thead", className, children);
}
exports.thead = thead;
function tbody(className, children) {
    return element("tbody", className, children);
}
exports.tbody = tbody;
function th(className, children) {
    return element("th", className, children);
}
exports.th = th;
function tr(className, children) {
    return element("tr", className, children);
}
exports.tr = tr;
function td(className, children) {
    return element("td", className, children);
}
exports.td = td;
function img(url) {
    var img = element("img", "", null);
    img.setAttribute("src", url);
    return img;
}
exports.img = img;
function input(type, className, value) {
    if (value === void 0) { value = ""; }
    var input = element("input", className, null);
    input.setAttribute("type", type);
    input.setAttribute("value", value);
    return input;
}
exports.input = input;
