// introcs/console/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ConsoleProvider_1 = require("./ConsoleProvider");
function print(s) {
    ConsoleProvider_1.default.instance().print(s);
}
exports.print = print;
function image(url) {
    ConsoleProvider_1.default.instance().image(url);
}
exports.image = image;
function promptNumber(prompt) {
    return ConsoleProvider_1.default.instance().promptNumber(prompt);
}
exports.promptNumber = promptNumber;
function promptString(prompt) {
    return ConsoleProvider_1.default.instance().promptString(prompt);
}
exports.promptString = promptString;
function promptBoolean(prompt) {
    return ConsoleProvider_1.default.instance().promptBoolean(prompt);
}
exports.promptBoolean = promptBoolean;
function csvToArray(prompt, classname) {
    return ConsoleProvider_1.default.instance().csvToArray(prompt, classname);
}
exports.csvToArray = csvToArray;
function csvToList(prompt, classname) {
    return ConsoleProvider_1.default.instance().csvToList(prompt, classname);
}
exports.csvToList = csvToList;
function clear() {
    ConsoleProvider_1.default.instance().clear();
}
exports.clear = clear;
function random(floor, ceiling) {
    return ConsoleProvider_1.default.instance().random(floor, ceiling);
}
exports.random = random;
function error(e) {
    ConsoleProvider_1.default.instance().error(e);
}
exports.error = error;
function setInterval(cb, duration) {
    return ConsoleProvider_1.default.instance().setInterval(cb, duration);
}
exports.setInterval = setInterval;
function setConsole(console) {
    ConsoleProvider_1.default.setConsole(console);
}
exports.setConsole = setConsole;
